File: java/src/main/java/io/hatake/core/FormatterRegistry.java
package io.hatake.core;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Resolves display format names to implementations. Built-ins and names are
 * shared with the other language editions.
 */
public final class FormatterRegistry {

    /** Turns a value into a display string using options. */
    @FunctionalInterface
    public interface Formatter {
        String format(Object value, Map<String, Object> options);
    }

    private final Map<String, Formatter> formatters = new HashMap<>();

    public FormatterRegistry() {
        formatters.putAll(builtins());
    }

    public FormatterRegistry(Map<String, Formatter> custom) {
        formatters.putAll(builtins());
        formatters.putAll(custom);
    }

    public String format(String name, Object value) {
        return format(name, value, Map.of());
    }

    public String format(String name, Object value, Map<String, Object> options) {
        Formatter f = formatters.get(name);
        return f == null ? str(value) : f.format(value, options);
    }

    public void register(String name, Formatter formatter) {
        formatters.put(name, formatter);
    }

    public boolean has(String name) {
        return formatters.containsKey(name);
    }

    private static String str(Object v) {
        return v == null ? "" : v.toString();
    }

    private static Double toNum(Object v) {
        if (v instanceof Number n) {
            return n.doubleValue();
        }
        if (v instanceof String s) {
            try {
                return Double.parseDouble(s.replace(",", ""));
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return null;
    }

    private static int intOpt(Map<String, Object> o, String key, int fallback) {
        return o.get(key) instanceof Number n ? n.intValue() : fallback;
    }

    private static String grouped(double value, int decimals) {
        String fixed = BigDecimal.valueOf(value)
                .setScale(decimals, RoundingMode.HALF_UP).toPlainString();
        int dot = fixed.indexOf('.');
        String intPart = dot < 0 ? fixed : fixed.substring(0, dot);
        String frac = dot < 0 ? null : fixed.substring(dot + 1);
        StringBuilder sb = new StringBuilder();
        int len = intPart.length();
        for (int i = 0; i < len; i++) {
            if (i > 0 && (len - i) % 3 == 0) {
                sb.append(',');
            }
            sb.append(intPart.charAt(i));
        }
        return frac == null ? sb.toString() : sb + "." + frac;
    }

    private static int[] ymd(Object v) {
        if (v instanceof LocalDate d) {
            return new int[] {d.getYear(), d.getMonthValue(), d.getDayOfMonth()};
        }
        if (v instanceof String s) {
            Matcher m = Pattern.compile("^(\\d{4})-(\\d{2})-(\\d{2})").matcher(s);
            if (m.find()) {
                return new int[] {
                    Integer.parseInt(m.group(1)),
                    Integer.parseInt(m.group(2)),
                    Integer.parseInt(m.group(3))
                };
            }
        }
        return null;
    }

    private static String two(int v) {
        return String.format("%02d", v);
    }

    private static String formatDatePattern(int y, int mo, int d, String pattern) {
        return pattern
                .replace("yyyy", String.format("%04d", y))
                .replace("MM", two(mo))
                .replace("dd", two(d))
                .replace("M", String.valueOf(mo))
                .replace("d", String.valueOf(d));
    }

    private static Map<String, Formatter> builtins() {
        Map<String, Formatter> m = new HashMap<>();
        m.put("currency", (value, o) -> {
            Double n = toNum(value);
            if (n == null) {
                return str(value);
            }
            String body = (o.get("symbol") instanceof String s ? s : "")
                    + grouped(Math.abs(n), intOpt(o, "decimals", 0));
            if (n < 0) {
                return switch (o.get("negative") instanceof String s ? s : "minus") {
                    case "triangle" -> "△" + body;
                    case "blackTriangle" -> "▲" + body;
                    case "paren" -> "(" + body + ")";
                    default -> "-" + body;
                };
            }
            return body;
        });
        m.put("percent", (value, o) -> {
            Double n = toNum(value);
            if (n == null) {
                return str(value);
            }
            double v = Boolean.TRUE.equals(o.get("ratio")) ? n * 100 : n;
            return grouped(v, intOpt(o, "decimals", 2)) + "%";
        });
        m.put("date", (value, o) -> {
            int[] d = ymd(value);
            if (d == null) {
                return str(value);
            }
            String pattern = o.get("pattern") instanceof String s ? s : "yyyy/MM/dd";
            return formatDatePattern(d[0], d[1], d[2], pattern);
        });
        m.put("wareki", (value, o) -> {
            int[] d = ymd(value);
            if (d == null) {
                return str(value);
            }
            Era.EraDate ed = Era.eraOfYmd(d[0], d[1], d[2]);
            if (ed == null) {
                return formatDatePattern(d[0], d[1], d[2], "yyyy/MM/dd");
            }
            if ("short".equals(o.get("style"))) {
                return ed.abbr() + ed.year() + "/" + two(d[1]) + "/" + two(d[2]);
            }
            String y = ed.year() == 1 ? "元" : String.valueOf(ed.year());
            return ed.name() + y + "年" + d[1] + "月" + d[2] + "日";
        });
        m.put("postal", (value, o) -> {
            String digits = str(value).replaceAll("[^0-9]", "");
            if (digits.length() != 7) {
                return str(value);
            }
            return digits.substring(0, 3) + "-" + digits.substring(3);
        });
        m.put("mask", (value, o) -> {
            String s = str(value);
            int keep = intOpt(o, "keep", 4);
            String ch = o.get("char") instanceof String c ? c : "*";
            if (s.length() <= keep) {
                return s;
            }
            return ch.repeat(s.length() - keep) + s.substring(s.length() - keep);
        });
        return m;
    }
}

File: flutter/packages/hatake_core/lib/src/format/formatter_registry.dart
import '../domain/era.dart';

/// A display formatter: turns a value into a display string using [options].
typedef Formatter = String Function(Object? value, Map<String, Object?> options);

num? _toNum(Object? v) {
  if (v is num) return v;
  if (v is String) return num.tryParse(v.replaceAll(',', ''));
  return null;
}

DateTime? _toDate(Object? v) {
  if (v is DateTime) return v;
  if (v is String && v.isNotEmpty) return DateTime.tryParse(v);
  return null;
}

int? _intOpt(Map<String, Object?> o, String key) => (o[key] as num?)?.toInt();

String _two(int v) => v.toString().padLeft(2, '0');

/// Groups the integer part with commas and applies fixed [decimals].
String _grouped(num value, int decimals) {
  final fixed = value.toStringAsFixed(decimals);
  final parts = fixed.split('.');
  final intPart = parts[0];
  final buf = StringBuffer();
  for (var i = 0; i < intPart.length; i++) {
    if (i > 0 && (intPart.length - i) % 3 == 0) buf.write(',');
    buf.write(intPart[i]);
  }
  return parts.length > 1 ? '$buf.${parts[1]}' : buf.toString();
}

String _formatDatePattern(DateTime d, String pattern) {
  return pattern
      .replaceAll('yyyy', d.year.toString().padLeft(4, '0'))
      .replaceAll('MM', _two(d.month))
      .replaceAll('dd', _two(d.day))
      .replaceAll('M', d.month.toString())
      .replaceAll('d', d.day.toString());
}

/// Built-in formatters. Names are shared across language editions.
final Map<String, Formatter> builtinFormatters = {
  // 金額: 1,234,567 / △1,234 / ▲1,234 / (1,234) / ¥1,234
  'currency': (value, o) {
    final n = _toNum(value);
    if (n == null) return value?.toString() ?? '';
    final decimals = _intOpt(o, 'decimals') ?? 0;
    final symbol = o['symbol'] as String? ?? '';
    final negative = o['negative'] as String? ?? 'minus';
    final body = symbol + _grouped(n.abs(), decimals);
    if (n < 0) {
      return switch (negative) {
        'triangle' => '△$body',
        'blackTriangle' => '▲$body',
        'paren' => '($body)',
        _ => '-$body',
      };
    }
    return body;
  },

  // パーセント: 12.34% / 12% / 12.3400%（ratio:true なら値を100倍）
  'percent': (value, o) {
    final n = _toNum(value);
    if (n == null) return value?.toString() ?? '';
    final decimals = _intOpt(o, 'decimals') ?? 2;
    final v = o['ratio'] == true ? n * 100 : n;
    return '${_grouped(v, decimals)}%';
  },

  // 日付: pattern で yyyy/MM/dd, yyyy-MM-dd, yyyy年M月d日, yyyyMMdd など
  'date': (value, o) {
    final d = _toDate(value);
    if (d == null) return value?.toString() ?? '';
    return _formatDatePattern(d, o['pattern'] as String? ?? 'yyyy/MM/dd');
  },

  // 和暦: 令和8年7月22日 (long) / R8/07/22 (short)。元号は eraOf と共有。
  'wareki': (value, o) {
    final d = _toDate(value);
    if (d == null) return value?.toString() ?? '';
    final ed = eraOf(d);
    if (ed == null) return _formatDatePattern(d, 'yyyy/MM/dd');
    if ((o['style'] as String? ?? 'long') == 'short') {
      return '${ed.abbr}${ed.year}/${_two(d.month)}/${_two(d.day)}';
    }
    final y = ed.year == 1 ? '元' : ed.year.toString();
    return '${ed.name}$y年${d.month}月${d.day}日';
  },

  // 郵便番号: 1234567 → 123-4567
  'postal': (value, o) {
    final digits = (value?.toString() ?? '').replaceAll(RegExp(r'[^0-9]'), '');
    if (digits.length != 7) return value?.toString() ?? '';
    return '${digits.substring(0, 3)}-${digits.substring(3)}';
  },

  // マスク: 末尾 keep 桁だけ残す → ******1234
  'mask': (value, o) {
    final s = value?.toString() ?? '';
    final keep = _intOpt(o, 'keep') ?? 4;
    final ch = o['char'] as String? ?? '*';
    if (s.length <= keep) return s;
    return ch * (s.length - keep) + s.substring(s.length - keep);
  },
};

/// Resolves format names to implementations. Extensible via [register].
class FormatterRegistry {
  final Map<String, Formatter> _formatters;

  /// アプリが足したフォーマッタの名前だけ（組み込みは除く）。
  List<String> get customKeys => [
        for (final key in _formatters.keys)
          if (!builtinFormatters.containsKey(key)) key,
      ]..sort();

  FormatterRegistry([Map<String, Formatter>? custom])
      : _formatters = {...builtinFormatters, if (custom != null) ...custom};

  /// Formats [value] with the formatter named [name]. Unknown names fall back
  /// to `toString()`.
  String format(String name, Object? value,
      [Map<String, Object?> options = const {}]) {
    final f = _formatters[name];
    return f == null ? (value?.toString() ?? '') : f(value, options);
  }

  void register(String name, Formatter formatter) =>
      _formatters[name] = formatter;

  bool has(String name) => _formatters.containsKey(name);
}
